### ISSUE TITLE
Silent detect in player.c  

### ISSUE THREAD

**Issue by juliandc1969** [mikebrady/shairport-sync#1000]
Hi Mike,
Been trying some workarounds for my SPS remote pause issue and until today I thought I was on to something. Using your commented code in player.c void *player_thread_func() I am able to detect null sample buffers after the 2 second drain period and this I'd use as silent period detection. After a remote pause I'd typically see the sample buffer drain out over 2 seconds and then for the next 5 seconds sample buffers full of nulls ( followed but session termination player_stop()  ).

Using your code I could detect the silent period and also when ( during a pause ) someone quickly press play again.

This has been going great for days, however today silent detection continuously failed. I tracked it down to the supposed null sample buffers ( being used for silent detection ) being filled with around  50% nulls.

Based on your commented code, this is a 2 second sample buffer draining snip-it after a pause:
zero 2 samples 353 blocks 229
zero 4 samples 352 blocks 230
zero 8 samples 351 blocks 231
zero 44 samples 352 blocks 232    // zero = nulls in buffer, samples = samples in buffer, blocks = checked sample buffers 
// The sample buffer drain is fine in both working and nonworking

The silent period after the above should look like: 
zero 352 samples 352 blocks 233
zero 353 samples 353 blocks 234
zero 353 samples 353 blocks 235
zero 352 samples 352 blocks 236  // nulls == samples

Instead now I get:
zero 186 samples 352 blocks 233
zero 181 samples 353 blocks 234
zero 159 samples 353 blocks 235
zero 198 samples 352 blocks 236

My code based on your code:
int co;
short *p = (short *)conn->outbuf;
int zero_sample_count=0;

for (co=0;co<play_samples;co++)
{
    if (*p==0)
    {
        zero_sample_count++;
    }
    p++;
}

Any idea why after days of working great it changes?  I have reverted all my changes, rebooted everything many times, still the same. 

Is the sample size what you'd expect? 

Thanks.

**Comment by mikebrady** [comment]
Thanks for the post. It's good news that the fix in 3.3.7d4 is working as well as it does.

I vaguely remember looking at the possibility of monitoring the audio itself to help decide what to do, and eventually concluded that it was chasing a chimera. 

The content of silence itself is even problematic, as it may well have dither – very low level white noise – added to it, so it won't always be a succession of samples of value zero.

Apart from everything else, the behaviour of different clients varies – whether the be iTunes, iOS, YouTube, etc. etc., so there is no constancy.

**Comment by juliandc1969** [comment]
Cheers! Interestingly there's always two "buffer types" I've noticed in play and pause/silence ( iPad ), why who knows; I should examine the buffer.. In brief this is what I see.

Type 1:
Playing - the audio buffer has a very low number of nulls
Paused/silence - the audio buffer is ~ 50% of nulls

Type 2:
Playing - the audio buffer has a null count of ~50% nulls
Paused/silence - the amount of nulls equal the audio buffer size

There's no specific thing I can do to encourage one type or the other, it can use one type for hours if not days, then just switch.   

I've added a fudge factor for both types so my silence detection seems happy. This is all because iPad YouTube returns very little in the way of AVRCP and there's no real way SPS side to know if things are truly playing of paused. When YouTube is minimised or if nothing plays for 60+ seconds then it closes the session.

I'll close the ticket as it's not a bug as such more just understanding SPS under the hood.

Thanks.
